[C10D] Implement new libuv backend for TCPStore.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105870

The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

kumpera committed:
[C10D] Implement new libuv backend for TCPStore.

The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera commented on test/distributed/test_store.py:
```diff
@@ -398,6 +401,17 @@ def test_multi_get(self):
         self.assertEqual(b"po", v0)
         self.assertEqual(b"tato", v1)
 
+class LibUvTCPStoreTest(TCPStoreTest):
```
good point, let me update those as well.

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp:
```diff
@@ -16,12 +17,1054 @@
+    return;
+  }
+
+  if (tryListen(true)) {
+    return;
+  }
+  uv_close((uv_handle_t*)&server, nullptr);
+  uv_run(&loop, UV_RUN_ONCE);
+
+  if (tryListen(false)) {
+    return;
+  }
+  uv_close((uv_handle_t*)&server, nullptr);
+  uv_run(&loop, UV_RUN_ONCE);
```
nit: can this be updated to be?
```C++
  if (tryListen(true) || tryListen(false)) {
    return;
  }
  uv_close((uv_handle_t*)&server, nullptr);
  uv_run(&loop, UV_RUN_ONCE);
```

I'm not familiar with the reason for calling uv_close and uv_run twice

kumpera replied:
This part of the code is really really fuggly.
We need to call close otherwise we leak the socket as tryListen creates a new one.
We need to call `uv_run` as operations are asynchronous and we can't reuse the handle memory until the uv loop processes and closes it.

H-Huang commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -3,6 +3,7 @@
 #include <torch/csrc/distributed/c10d/TCPStoreBackend.hpp>
 #include <torch/csrc/distributed/c10d/logging.h>
 
+#include <distributed/c10d/Backend.hpp>
```
is this needed?

kumpera replied:
no, this is garbage.

H-Huang commented on torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp:
```diff
@@ -16,12 +17,1054 @@
+  TORCH_CHECK(false, "failed to init store, no bind possible");
+}
+
+bool LibUVStoreDaemon::tryListen(bool use_ipv6) {
```
Im confused since each tcpstore instance calls into this function? Doesn't that mean that each instance is goign to try to bind to that address when it should just be the master?

kumpera replied:
`LibUVStoreDaemon` only implements the server part of the Store interface. The check you mention is done in TCPStore.cpp and is shared between libuv and the existing store implementation.

The client part of TCPStore is shared between the two backends.

H-Huang reviewed (commented):
Tried my best to review, but I still have gaps on the libuv APIs that I need to understand.

Let's schedule a meeting and go over the `TCPStoreLibUvBackend.cpp`? I think that would help to speed the review

kumpera commented on torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp:
```diff
@@ -16,12 +17,1054 @@
+    write_value<uint64_t>(val.size());
+    data.insert(data.end(), val.data(), val.data() + val.size());
+  }
+  void send() {
+    auto wd = new WriterPayload(std::move(data), handle);
+    if (!wd->send())
+      delete wd;
+  }
```
If `send` returns true, it means that the write was queued and WriterPayload must stick around until the write callback is invoked.

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Implement new libuv backend for TCPStore."


The new backend is currently under a flag 'use_libuv' in TCPStore constructor
to reduce the impact on existing users as we test it.

[ghstack-poisoned]

H-Huang reviewed (approved):
Updates look good, thank you for adding! It will be awesome to see benchmarks once checked in

kumpera commented:
@pytorchmergebot merge -f "failures unrelated"

huydhn commented:
@pytorchbot revert -m 'I think the distributed failure is related as this is now failing in trunk' -c ignoredsignal

ZainRizvi commented:
Hi @kumpera, could you please help us understand why you thought the failure was unrelated? (Not blame gaming, trying to see how we can improve the tooling to build more confidence in the results)

huydhn commented:
@pytorchbot drci

kumpera commented:
@huydhn the PR adds an optional component that is not used outside its own tests.
